### PR TITLE
State Build implementation for File Copy based replication

### PR DESCRIPTION
**Pull request by piyujai**
This PR handles initializing BlobStore with Log, Index, BlobStoreCompactor, RemoteTokenTracker, etc post filecopy is completed in the filecopy based replication. 

[Bootstrap State Build Design Doc](https://docs.google.com/document/d/1pb8KuA2vT8L5FrxCobuT_IHWMtnWnNF0uKPHtqcMXN0/edit?tab=t.0)

Key Changes:
1. **buildStateForFileCopy**: New method to build in-memory state when a file copy completes under file copy based replication. It throws an exception if ReplicaId is null and adjusts disk space on failure.
2. **addBlobStoreForFileCopy**: Added to DiskManager and StorageManager for adding BlobStores during file copy replication. It includes initializing of BlobStore with Log, Index, BlobStoreCompactor and RemoteTokenTracker, etc. It includes error handling for failed store creation.
3. **Error Handling**: Enhanced logging and handling for failed store additions and bootstrap replica issues.

**Test Coverage:**
1. Added tests for various file copy scenarios, including successful and failed store additions and handling null ReplicaId, successful state build post blob writes to a new store, etc.
2. Testing with sealed log files with setting of active segments and rollover of files is working correctly in local testing.

### CODE REVIEW

File: ambry-store/src/main/java/com/github/ambry/store/StorageManager.java [hunk 1]
```diff
@@ -530,11 +552,45 @@ public boolean addBlobStore(ReplicaId replica) {
...
+    if (replica == null) {
+      logger.error("ReplicaId is null");
+      throw new StateTransitionException("ReplicaId null is not found in clustermap for " + currentNode, ReplicaNotFound);
+    }
+    PartitionId partitionId = replica.getPartitionId();
+
+    if (!addBlobStoreForFileCopy(replica)){
+      // We have decreased the available disk space in HelixClusterManager#getDiskForBootstrapReplica. Increase it
+      // back since addition of store failed.
+      replica.getDiskId().increaseAvailableSpaceInBytes(replica.getCapacityInBytes());
```
Comment (aga9900) [inline ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:571]: We should have cleanup before decreasing the space on disk.

Comment (aga9900) [reply ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:571]: Or, we should just throw exception and let Helix take care of exception and wait for downward transition.

Comment (piyujai) [reply ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:571]: The disk space is already reduced in pre-filecopy steps since diskAllocator reserves the space. This increase in space is to undo that reservation done earlier.

**Review by DevenAhluwalia — DISMISSED** [review]

File: ambry-store/src/main/java/com/github/ambry/store/StorageManager.java [hunk 2]
```diff
@@ -514,6 +514,28 @@ DiskManager addDisk(DiskId diskId) {
...
   }
 
+  /**
+   * Add a new store to the storage manager for file copy based replication post filecopy is completed.
+   * @param replica the {@link ReplicaId} of the {@link Store} which would be added.
+   * @return
+   */
+  @Override
+  public boolean addBlobStoreForFileCopy(ReplicaId replica) {
+    if (partitionToDiskManager.containsKey(replica.getPartitionId())) {
```
Comment (aga9900) [inline ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:524]: This will get added as part of file store, so it might occur that the this always returns true.

Comment (aga9900) [reply ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:524]: We should instead check below map for existence of store in disk manager.
  private final ConcurrentHashMap<PartitionId, BlobStore> stores = new ConcurrentHashMap<>();

File: ambry-store/src/main/java/com/github/ambry/store/StorageManager.java [hunk 3]
```diff
@@ -530,11 +552,45 @@ public boolean addBlobStore(ReplicaId replica) {
...
+      // back since addition of store failed.
+      replica.getDiskId().increaseAvailableSpaceInBytes(replica.getCapacityInBytes());
+      if (!clusterMap.isDataNodeInFullAutoMode(currentNode)) {
+        logger.error("Failed to add store for replica {} into storage manager", partitionId.getId());
+        throw new StateTransitionException("Failed to add store for replica " + partitionId.getId() + " into storage manager",
+            ReplicaOperationFailure);
+      } else {
+        logger.info("Failed to add store for replica {} at location {}. Retrying bootstrapping replica at different location",
+            partitionId.getId(), replica.getReplicaPath());
+        tryRemoveFailedBootstrapBlobStore(replica);
```
Comment (aga9900) [inline ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:579]: We need to add StateTransitiojException here and put replica in error state.

File: ambry-store/src/main/java/com/github/ambry/store/DiskManager.java [hunk 4]
```diff
@@ -419,6 +419,44 @@ boolean controlCompactionForBlobStore(PartitionId id, boolean enabled) {
...
+    boolean succeed = false;
+    try {
+      if (!running) {
+        logger.error("Failed to add {} because disk manager is not running", replica.getPartitionId());
+      } else {
+        // Directory is already created in prefilecopy steps. So no directory cleanup required.
+        BlobStore store = new BlobStore(replica, storeConfig, scheduler, longLivedTaskScheduler, this, diskIOScheduler,
+            diskSpaceAllocator, storeMainMetrics, storeUnderCompactionMetrics, keyFactory, recovery, hardDelete,
+            replicaStatusDelegates, time, accountService, null, indexPersistScheduler);
+        store.start();
```
Comment (justinlin-linkedin) [inline ambry-store/src/main/java/com/github/ambry/store/DiskManager.java:439]: BlobStore.start would allocate first log segment file 0_0_log, which is not needed for FileCopy.

File: ambry-store/src/main/java/com/github/ambry/store/DiskManager.java [hunk 5]
```diff
@@ -419,6 +419,44 @@ boolean controlCompactionForBlobStore(PartitionId id, boolean enabled) {
...
+        // add store into CompactionManager
+        compactionManager.addBlobStore(store);
+        // add new created store into in-memory data structures.
+        stores.put(replica.getPartitionId(), store);
+        partitionToReplicaMap.put(replica.getPartitionId(), replica);
+        // create a bootstrap-in-progress file to distinguish it from regular stores (the file will be checked during
+        // BOOTSTRAP -> STANDBY transition)
+        createBootstrapFileIfAbsent(replica);
```
Comment (justinlin-linkedin) [inline ambry-store/src/main/java/com/github/ambry/store/DiskManager.java:447]: why do we want to create a bootstrapFile, we should create a FileCopy file here right?

File: ambry-store/src/main/java/com/github/ambry/store/StorageManager.java [hunk 6]
```diff
@@ -530,11 +552,45 @@ public boolean addBlobStore(ReplicaId replica) {
...
-  public void buildStateForFileCopy(String partitionName){
-    // no-op
+  public void buildStateForFileCopy(ReplicaId replica){
+    if (replica == null) {
+      logger.error("ReplicaId is null");
+      throw new StateTransitionException("ReplicaId null is not found in clustermap for " + currentNode, ReplicaNotFound);
+    }
+    PartitionId partitionId = replica.getPartitionId();
+
+    if (!addBlobStoreForFileCopy(replica)){
```
Comment (justinlin-linkedin) [inline ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:568]: I am confused, I thought we are going to addBlobStoreForFileCopy before the FileCopy finishes all the log files?

File: ambry-store/src/main/java/com/github/ambry/store/DiskManager.java (hunk 4 above)
Comment (piyujai) [reply ambry-store/src/main/java/com/github/ambry/store/DiskManager.java:439]: BlobStore.start() will init 0_0_log only if no files present in the partition directory. For our case, since filecopy will have sealed files in the partition directory, it will make the last file as active segment, which is the expected behaviour in State Build

File: ambry-store/src/main/java/com/github/ambry/store/DiskManager.java (hunk 5 above)
Comment (piyujai) [reply ambry-store/src/main/java/com/github/ambry/store/DiskManager.java:447]: 2 things here - 
1. bootstrapFile will be required here since we intend to use the existing replication flow for active segments and BOOTSTRAP->STANDBY transition checks for this file.
2. creating a FILE_COPY_IN_PROGRESS file will be taken up by FileCopyManager before the stateBuild gets triggered.

File: ambry-store/src/main/java/com/github/ambry/store/StorageManager.java [hunk 7]
```diff
@@ -514,6 +514,28 @@ DiskManager addDisk(DiskId diskId) {
     });
   }
 
+  /**
+   * Add a new store to the storage manager for file copy based replication post filecopy is completed.
+   * @param replica the {@link ReplicaId} of the {@link Store} which would be added.
+   * @return
+   */
+  @Override
+  public boolean addBlobStoreForFileCopy(ReplicaId replica) {
```
[the remark this answers is not in the document]
Comment (piyujai) [reply ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:526]: buildStateForFileCopy will be required if later we have a different implementation for StoreManager, similar to current usecase. Hence we will need it in the interface.

File: ambry-store/src/main/java/com/github/ambry/store/StorageManager.java (hunk 2 above)
Comment (piyujai) [reply ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:527]: Yes, diskmanager will be started already in pre-filecopy steps. Instead, I have put the check of partitionToDiskManager not having the partitionId key and fetching the diskManager from this map itself. Checking of store's existence is handled in the addBlobStoreForFileCopy call.

File: ambry-store/src/main/java/com/github/ambry/store/StorageManager.java (hunk 6 above)
Comment (piyujai) [reply ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:571]: @justinlin-linkedin As per the design of StateBuild and Bootstrap FileCopy, we have the flow where we we don't addBlobStore before filecopy, but only once state building is triggered. The trigger point for state build will handle latest log and index segment marking active and starting the BlobStore.

File: ambry-store/src/main/java/com/github/ambry/store/StorageManager.java [hunk 8]
```diff
@@ -530,11 +552,45 @@ public boolean addBlobStore(ReplicaId replica) {
...
+    if (!addBlobStoreForFileCopy(replica)){
+      // We have decreased the available disk space in HelixClusterManager#getDiskForBootstrapReplica. Increase it
+      // back since addition of store failed.
+      replica.getDiskId().increaseAvailableSpaceInBytes(replica.getCapacityInBytes());
+      if (!clusterMap.isDataNodeInFullAutoMode(currentNode)) {
+        logger.error("Failed to add store for replica {} into storage manager", partitionId.getId());
+        throw new StateTransitionException("Failed to add store for replica " + partitionId.getId() + " into storage manager",
+            ReplicaOperationFailure);
+      } else {
+        logger.info("Failed to add store for replica {} at location {}. Retrying bootstrapping replica at different location",
```
[the remark this answers is not in the document]
Comment (piyujai) [reply ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:580]: This is not retry, Inside this method, we have cleanup steps in this util.

File: ambry-store/src/test/java/com/github/ambry/store/StorageManagerTest.java [hunk 9]
```diff
@@ -269,6 +270,131 @@ public void scheduleAndControlCompactionTest() throws Exception {
...
+    storageManager.start();
+    // test add store onto a new disk, which should succeed
+    assertTrue("Add new store should succeed", storageManager.addBlobStore(newPartition.getReplicaIds().get(0)));
+    assertNotNull("The store shouldn't be null because new store is successfully added",
+        storageManager.getStore(newPartition, false));
+    DiskManager dm = storageManager.getDiskManager(newPartition);
+    Store store = dm.getStore(newPartition, false);
+    MockId id1 = addRandomBlobToStore(store, 100, Utils.Infinite_Time);
+    MockId id2 = addRandomBlobToStore(store, 200, Utils.Infinite_Time);
+    // Create storage manager again to create disk managers again
```
[the remark this answers is not in the document]
Comment (piyujai) [reply ambry-store/src/test/java/com/github/ambry/store/StorageManagerTest.java:326]: We need to shut it down to ensure that BlobStores are shut down and state build is able to start the BlobStores again. This is to test the flow where existing blobs are written to disk and then building state is attempted by adding a store via stateBuild flow.

File: ambry-store/src/test/java/com/github/ambry/store/StorageManagerTest.java [hunk 10]
```diff
@@ -269,6 +270,131 @@ public void scheduleAndControlCompactionTest() throws Exception {
...
+    storageManager.shutdown();
+    storageManager = createStorageManager(localNode, metricRegistry, null);
+    storageManager.start();
+    storageManager.buildStateForFileCopy(newPartition.getReplicaIds().get(0));
+    dm = storageManager.getDiskManager(newPartition);
+    store = dm.getStore(newPartition, false);
+    assertNotNull(store.get(Collections.singletonList(id1), EnumSet.noneOf(StoreGetOptions.class)));
```
[the remark this answers is not in the document]
Comment (piyujai) [reply ambry-store/src/test/java/com/github/ambry/store/StorageManagerTest.java:333]: This assertion means that blob with id1 exists. We have passed none in StoreGetOptions since we aren't concerned with expired of deleted blobs currently.

[the remark this answers is not in the document]
Comment (piyujai) [reply ambry-store/src/test/java/com/github/ambry/store/StorageManagerTest.java:333]: @DevenAhluwalia We can only verify the existence of blob with id since we are writing random content into blobs from the above util addRandomBlobToStore

File: ambry-store/src/main/java/com/github/ambry/store/StorageManager.java [hunk 11]
```diff
@@ -517,6 +517,28 @@ DiskManager addDisk(DiskId diskId) {
...
+    if (partitionToDiskManager.containsKey(replica.getPartitionId())) {
+      logger.info("{} already exists in storage manager, rejecting adding store request", replica.getPartitionId());
+      return false;
+    }
+    DiskManager diskManager = addDisk(replica.getDiskId());
+    if (diskManager == null || !diskManager.addBlobStoreForFileCopy(replica)) {
+      logger.error("Failed to add new store into DiskManager");
+      return false;
+    }
+    partitionToDiskManager.put(replica.getPartitionId(), diskManager);
```
Comment (aga9900) [inline ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:536]: this should already be present when we add the Blob Store after file copy.

File: ambry-store/src/test/java/com/github/ambry/store/StorageManagerTest.java [hunk 12]
```diff
@@ -269,6 +270,131 @@ public void scheduleAndControlCompactionTest() throws Exception {
...
+        new MockPartitionId(newPartitionId, MockClusterMap.DEFAULT_PARTITION_CLASS, clusterMap.getDataNodes(), newMountPathIndex);
+    StorageManager storageManager = createStorageManager(localNode, metricRegistry, null);
+    storageManager.start();
+    // test add store onto a new disk, which should succeed
+    assertTrue("Add new store should succeed", storageManager.addBlobStore(newPartition.getReplicaIds().get(0)));
+    assertNotNull("The store shouldn't be null because new store is successfully added",
+        storageManager.getStore(newPartition, false));
+    // This should fail since the newPartition already has store started.
+    assertFalse("Add store which is already existing should fail", storageManager.addBlobStoreForFileCopy(newPartition.getReplicaIds().get(0)));
+    storageManager.buildStateForFileCopy(null);
```
[the remark this answers is not in the document]
Comment (piyujai) [reply ambry-store/src/test/java/com/github/ambry/store/StorageManagerTest.java:394]: I have refactored all 3 tests to use common utils.

File: ambry-store/src/main/java/com/github/ambry/store/DiskManager.java [hunk 13]
```diff
@@ -420,6 +420,44 @@ boolean controlCompactionForBlobStore(PartitionId id, boolean enabled) {
...
+        // Directory is already created in prefilecopy steps. So no directory cleanup required.
+        BlobStore store = new BlobStore(replica, storeConfig, scheduler, longLivedTaskScheduler, this, diskIOScheduler,
+            diskSpaceAllocator, storeMainMetrics, storeUnderCompactionMetrics, keyFactory, recovery, hardDelete,
+            replicaStatusDelegates, time, accountService, null, indexPersistScheduler);
+        store.start();
+        // add store into CompactionManager
+        compactionManager.addBlobStore(store);
+        // add new created store into in-memory data structures.
+        stores.put(replica.getPartitionId(), store);
+        partitionToReplicaMap.put(replica.getPartitionId(), replica);
```
Comment (aga9900) [inline ambry-store/src/main/java/com/github/ambry/store/DiskManager.java:445]: As the name suggest, we should be adding an entry to this map much sooner than the file copy finishes.

File: ambry-store/src/main/java/com/github/ambry/store/StorageManager.java [hunk 14]
```diff
@@ -536,6 +536,29 @@ DiskManager addDisk(DiskId diskId) {
...
+      return false;
+    }
+    // We don't require addDisk since DiskManager is already started during initialization of StorageManager as part
+    // of prefilecopy steps. We will fetch it from partitionToDiskManager map.
+    DiskManager diskManager = partitionToDiskManager.get(replica.getPartitionId());
+    if (diskManager == null || !diskManager.addBlobStoreForFileCopy(replica)) {
+      logger.error("Failed to add new store into DiskManager");
+      return false;
+    }
+    partitionNameToReplicaId.put(replica.getPartitionId().toPathString(), replica);
```
Comment (aga9900) [inline ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:557]: Same as this: https://github.com/linkedin/ambry/pull/2954/files#r1894565174

File: ambry-store/src/test/java/com/github/ambry/store/StorageManagerTest.java [hunk 15]
```diff
@@ -269,6 +270,131 @@ public void scheduleAndControlCompactionTest() throws Exception {
...
 
+  /**
+   * Helper util to add blobs to a given Store
+   * @param store store to add blob to.
+   * @param size size in bytes of the randomized blob
+   * @param expiresAtMs expiry in milliseconds to be set for the blob
+   * @return {@link MockId} the mock id of the blob added
+   * @throws StoreException
+   */
+  public MockId addRandomBlobToStore(Store store, long size, long expiresAtMs) throws StoreException {
```
Comment (aga9900) [inline ambry-store/src/test/java/com/github/ambry/store/StorageManagerTest.java:281]: Can we add test for blob removal as well, just to make sure that the new addition is compatible with the removal flow.

**Review by aga9900 — CHANGES_REQUESTED** [review]
Added couple of comments.

File: ambry-store/src/main/java/com/github/ambry/store/StorageManager.java (hunk 1 above)
Comment (aga9900) [reply ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:598]: 1. Shouldn't we reduce the space on disk after removing the directory based on success or failure of tryRemoveFailedBootstrapBlobStore().
2. Update state in Helix as well.
Or
We can delegate the whole responsibility to the downward state transition of clean up in Helix.

File: ambry-store/src/test/java/com/github/ambry/store/StorageManagerTest.java (hunk 15 above)
Comment (aga9900) [reply ambry-store/src/test/java/com/github/ambry/store/StorageManagerTest.java:281]: More tests for all negative conditions for all exceptions thrown at Disk Manager, Storage Manager and Blob Store Addition + removal.

Comment (piyujai) [reply ambry-store/src/test/java/com/github/ambry/store/StorageManagerTest.java:281]: @aga9900 I will be adding a TODO for additional tests since we will have new tests once FC and prefilecopy flow done as well.

File: ambry-store/src/main/java/com/github/ambry/store/StorageManager.java (hunk 1 above)
Comment (piyujai) [reply ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:598]: 1. Before FIleCopy, we reduce the space available. Since we fail out, we have to increase the space now.

File: ambry-store/src/main/java/com/github/ambry/store/DiskManager.java (hunk 13 above)
Comment (piyujai) [reply ambry-store/src/main/java/com/github/ambry/store/DiskManager.java]: Done, we should handle it in prefilecopy. Please make a note.

File: ambry-store/src/main/java/com/github/ambry/store/StorageManager.java (hunk 14 above)
Comment (piyujai) [reply ambry-store/src/main/java/com/github/ambry/store/StorageManager.java]: Done, we should handle it in prefilecopy. Please make a note.

**Review by DevenAhluwalia — APPROVED** [review]
🚢
